utils/man/man1/powercap-set.1
.TH "POWERCAP\-SET" "1" "2021-12-12" "powercap" "powercap\-set"
.SH "NAME"
.LP
powercap\-set \- manage power capping devices with the Linux power capping
framework
.SH "SYNPOSIS"
.LP
\fBpowercap\-set\fP \fINAME\fP [\fIOPTION\fP]...
.SH "DESCRIPTION"
.LP
Sets configurations for a powercap control type.
.LP
The control type \fINAME\fP must not be empty or contain a '.' or '/'.
.SH "OPTIONS"
.LP
.TP
\fB\-h,\fR \fB\-\-help\fR
Prints the help screen
.TP
\fB\-p,\fR \fB\-\-control\-type\fR=\fINAME\fP
Deprecated, provide \fINAME\fP as the first positional argument instead.
.TP
\fB\-z,\fR \fB\-\-zone=\fR\fIZONE(S)\fP
The zone/subzone numbers in the control type's powercap tree.
Separate zones/subzones with a colon.
E.g., for zone 0, subzone 2:
.br
\fB\-z 0:2\fP
.TP
\fB\-c,\fR \fB\-\-constraint\fR=\fICONSTRAINT\fP
The constraint number (none by default)
.TP
\fB\-E,\fR \fB\-\-enabled=1|0\fR
Enable/disable a control type
.LP
The following zone-level arguments may be used together and require
\-z/\-\-zone:
.TP
\fB\-j,\fR \fB\-\-z\-energy\fR
Reset zone energy counter
.TP
\fB\-e,\fR \fB\-\-z\-enabled=1|0\fR
Enable/disable a zone
.LP
The following constraint-level arguments may be used together and require
\-z/\-\-zone and \-c/\-\-constraint:
.TP
\fB\-l,\fR \fB\-\-c\-power\-limit=UW\fR
Set constraint power limit
.TP
\fB\-s,\fR \fB\-\-c\-time\-window=US\fR
Set constraint time window
.SH "EXAMPLES"
.TP
\fBpowercap\-set intel\-rapl \-z 0 \-e 1\fP
Enable zone 0 for the \fIintel\-rapl\fR control type.
.TP
\fBpowercap\-set intel\-rapl \-z 0:1 \-e 1\fP
Enable zone 0, subzone 1 for the \fIintel\-rapl\fR control type.
.TP
\fBpowercap\-set intel\-rapl \-z 0 \-c 1 \-l 25000000\fP
Set a power cap of 25 Watts (25000000 uW) on zone 0, constraint 1 for the
\fIintel\-rapl\fR control type.
.TP
\fBpowercap\-set intel\-rapl \-z 1:0 \-c 0 \-l 15000000 \-s 976\fP
Set a power cap of 15 Watts (15000000 uW) and time window of 976
microseconds on zone 1, subzone 0, constraint 0 for the \fIintel\-rapl\fR
control type.
.SH "REMARKS"
.LP
Administrative (root) privileges are usually needed to use
\fBpowercap\-set\fR.
.LP
Setting constraint power cap and/or time limit values may not enable or
disable a zone/subzone - set the \-e/-\-z\-enabled flag explicitly.
.LP
The kernel may round power limits and time windows to discrete values
supported by the hardware, e.g., 28 seconds might round to 27983872
microseconds.
.LP
Power units: microwatts (uW)
.br
Time units: microseconds (us)
.SH "BUGS"
.LP
Report bugs upstream at <https://github.com/powercap/powercap>
.SH "FILES"
.nf
\fI/sys/devices/virtual/powercap/*\fP
.nf
\fI/sys/class/powercap/*\fP
.fi
.SH "AUTHORS"
.nf
Connor Imes <connor.k.imes@gmail.com>
.fi
.SH "SEE ALSO"
.BR powercap\-info (1)

RELEASES.md
# Release Notes

## [Unreleased]

### Added

* powercap-info: control type name no longer required - will print info for all control types if not set


## [v0.5.0] - 2021-12-09

### Added

* CMake installation components

### Changed

* Increased minimum CMake version from 3.1 to 3.12
* CMake package config now exports binaries in a 'Utils' component rather than parent package


## [v0.4.0] - 2021-04-25

### Fixed

* powercap.h: header documentation for 'powercap_*_file_get_name' functions' return behavior
* rapl-{info,set}: deprecation messages that specified non-existent alternatives

### Added

* Functions to support opening files in powercap.h:
  * 'powercap_control_type_file_open'
  * 'powercap_zone_file_open'
  * 'powercap_constraint_file_open'
* CMake helper to find powercap without relying on pkg-config

### Changed

* Binaries powercap-{info,set} now accept the control type argument without requiring a preceding -p flag
* Increased minimum CMake version from 2.8.12 to 3.1

### Deprecated

* Flag '-p/--control-type' in powercap-{info,set} - specify control type name as the first positional argument instead
* Add compiler deprecation attributes to deprecated functions


## [v0.3.1] - 2020-11-08

### Fixed

* [#6] powercap-common-test:test_snprintf_base_path fails on ppc64el architecture


## [v0.3.0] - 2020-08-15

### Added

* Support for top-level control type
  * Struct 'powercap_control_type', enum 'powercap_control_type_file', and associated functions in powercap.h:
    * 'powercap_control_type_get_enabled'
    * 'powercap_control_type_set_enabled'
    * 'powercap_control_type_file_get_name'
  * Functions in powercap-sysfs.h:
    * 'powercap_sysfs_control_type_get_enabled'
    * 'powercap_sysfs_control_type_set_enabled'
  * Functions in powercap-rapl.h:
    * 'powercap_rapl_control_is_supported'
    * 'powercap_rapl_control_is_enabled'
    * 'powercap_rapl_control_set_enabled'
  * Argument '--enabled' to powercap-{info,set} for getting/setting control type enabled field
* Function 'powercap_rapl_get_num_instances' in powercap-rapl.h (supersedes 'powercap_rapl_get_num_packages')
* Argument '--nconstraints' to powercap-info for getting the number of zone constraints

### Changed

* Increased minimum CMake version from 2.8.5 to 2.8.12 to support target_compile_definitions
* Updated 'powercap-{info,set}' man pages
* Argument '--zone' for powercap-set no longer required due to introduction of '--enabled' argument
* Disabled logging by default (in general, libraries shouldn't print output unless requested)

### Deprecated

* Binaries rapl-{info,set} - use powercap-{info,set} instead
* Interface powercap-rapl-sysfs.h - use powercap-sysfs.h directly instead
* Function 'powercap_rapl_get_num_packages' in powercap-rapl.h - use 'powercap_rapl_get_num_instances' instead

### Fixed

* Fixed [#4]: powercap-rapl now checks if parent zone is PACKAGE or PSYS


## [v0.2.0] - 2019-12-03

### Added

* This RELEASES.md file
* Multiarch support (use GNU standard installation directories)
* Additional documentation in README
* Function 'rapl_sysfs_zone_exists' in powercap-rapl-sysfs.h
* Command line long option '--zone' to rapl-info and rapl-set
* Functions to powercap-rapl-sysfs.h allowed by the powercap API that RAPL doesn't support (but could in the future)
* Options to rapl-info and rapl-set to support optional powercap features not currently implemented in RAPL

### Changed

* Increased minimum CMake version from 2.8 to 2.8.5 to support GNUInstallDirs
* More pedantic man page source formatting
* RAPL sysfs 'pkg' parameter names changed to 'zone' to improve generality
* RAPL 'package' parameter names changed to 'id' to improve generality
* On failure, binaries now exit with positive error codes instead of negative values

### Deprecated

* Function 'rapl_sysfs_pkg_exists' in powercap-rapl-sysfs.h - use 'rapl_sysfs_zone_exists' instead
* Function 'rapl_sysfs_sz_exists' in powercap-rapl-sysfs.h - use 'rapl_sysfs_zone_exists' instead
* Command line long option '--package' for rapl-info and rapl-set - use '--zone' instead

### Removed

* Removed private symbol exports in shared object library (already patched in Debian)
* Undocumented (and unintended) --package long option from powercap-set

### Fixed

* Fixed [#1]: Kernel uses hexadecimal numbers in directory paths, not decimal


## [v0.1.1] - 2017-09-21

### Added

* Added stateless interfaces to sysfs
* Added powercap-info, powercap-set, rapl-info, and rapl-set binaries and man pages
* Added VERSION and SOVERSION to shared object libraries

### Changed

* Update license to use author as copyright holder


## v0.1.0 - 2017-06-09

### Added

* Initial public release

[Unreleased]: https://github.com/powercap/powercap/compare/v0.5.0...HEAD
[v0.5.0]: https://github.com/powercap/powercap/compare/v0.4.0...v0.5.0
[v0.4.0]: https://github.com/powercap/powercap/compare/v0.3.1...v0.4.0
[v0.3.1]: https://github.com/powercap/powercap/compare/v0.3.0...v0.3.1
[v0.3.0]: https://github.com/powercap/powercap/compare/v0.2.0...v0.3.0
[v0.2.0]: https://github.com/powercap/powercap/compare/v0.1.1...v0.2.0
[v0.1.1]: https://github.com/powercap/powercap/compare/v0.1.0...v0.1.1
[#6]: https://github.com/powercap/powercap/issues/6
[#4]: https://github.com/powercap/powercap/issues/4
[#1]: https://github.com/powercap/powercap/issues/1


utils/powercap-set.c
/*
 * SPDX-License-Identifier: BSD-3-Clause
 *
 * Set powercap values.
 *
 * @author Connor Imes
 * @date 2017-08-24
 */
#include <errno.h>
#include <getopt.h>
#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include "powercap-sysfs.h"
#include "util-common.h"

static const char short_options[] = "-hp:z:c:E:je:l:s:";

static const struct option long_options[] = {
  {"help",                no_argument,        NULL, 'h'},
  {"control-type",        required_argument,  NULL, 'p'},
  {"zone",                required_argument,  NULL, 'z'},
  {"constraint",          required_argument,  NULL, 'c'},
  {"enabled",             required_argument,  NULL, 'E'},
  {"z-energy",            no_argument      ,  NULL, 'j'},
  {"z-enabled",           required_argument,  NULL, 'e'},
  {"c-power-limit",       required_argument,  NULL, 'l'},
  {"c-time-window",       required_argument,  NULL, 's'},
  {0, 0, 0, 0}
};

static void print_usage(void) {
  printf("Usage: powercap-set NAME [OPTION]...\n\n");
  printf("Sets configurations for a powercap control type.\n");
  printf("The control type NAME must not be empty or contain a '.' or '/'.\n\n");
  printf("Options:\n");
  printf("  -h, --help                   Print this message and exit\n");
  printf("  -p, --control-type=NAME      Deprecated, provide NAME as the first\n");
  printf("                               positional argument instead\n");
  printf("  -z, --zone=ZONE(S)           The zone/subzone numbers in the control type's\n");
  printf("                               powercap tree\n");
  printf("                               Separate zones/subzones with a colon\n");
  printf("                               E.g., for zone 0, subzone 2: \"-z 0:2\"\n");
  printf("  -c, --constraint=CONSTRAINT  The constraint number (none by default)\n");
  printf("  -E, --enabled=1|0            Enable/disable the control type\n");
  printf("The following zone-level arguments may be used together and require -z/--zone:\n");
  printf("  -j, --z-energy               Reset zone energy counter\n");
  printf("  -e, --z-enabled=1|0          Enable/disable a zone\n");
  printf("The following constraint-level arguments may be used together and require -z/--zone and -c/--constraint:\n");
  printf("  -l, --c-power-limit=UW       Set constraint power limit\n");
  printf("  -s, --c-time-window=US       Set constraint time window\n");
  printf("\nPower units: microwatts (uW)\n");
  printf("Time units: microseconds (us)\n");
}

static void print_common_help(void) {
  printf("Considerations for common errors:\n");
  printf("- Ensure that the control type exists, which may require loading a kernel module\n");
  printf("- Ensure that you run with administrative (super-user) privileges\n");
  printf("- Enabling/disabling a control type is an optional feature not supported by all control types\n");
  printf("- Resetting a zone energy counter is an optional powercap feature not supported by all control types\n");
}

int main(int argc, char** argv) {
  const char* control_type = NULL;
  u32_param ct_enabled = {0, 0};
  uint32_t zones[MAX_ZONE_DEPTH] = { 0 };
  uint32_t depth = 0;
  u32_param constraint = {0, 0};
  int reset_energy = 0;
  u32_param enabled = {0, 0};
  u64_param power_limit = {0, 0};
  u64_param time_window = {0, 0};
  int is_set_zone = 0;
  int is_set_constraint = 0;
  int c;
  int cont = 1;
  int ret = 0;

  /* Parse command-line arguments */
  while (cont) {
    c = getopt_long(argc, argv, short_options, long_options, NULL);
    switch (c) {
    case -1:
      cont = 0;
      break;
    case 'h':
      print_usage();
      return EXIT_SUCCESS;
    case 1:
    case 'p':
      if (control_type) {
        cont = 0;
        ret = -EINVAL;
      }
      control_type = optarg;
      break;
    case 'E':
      ret = set_u32_param(&ct_enabled, optarg, &cont);
      break;
    case 'z':
      ret = parse_zones(optarg, zones, MAX_ZONE_DEPTH, &depth, &cont);
      break;
    case 'c':
      ret = set_u32_param(&constraint, optarg, &cont);
      break;
    case 'j':
      if (reset_energy) {
        cont = 0;
        ret = -EINVAL;
      }
      reset_energy = 1;
      is_set_zone = 1;
      break;
    case 'e':
      ret = set_u32_param(&enabled, optarg, &cont);
      is_set_zone = 1;
      break;
    case 'l':
      ret = set_u64_param(&power_limit, optarg, &cont);
      is_set_constraint = 1;
      break;
    case 's':
      ret = set_u64_param(&time_window, optarg, &cont);
      is_set_constraint = 1;
      break;
    case '?':
    default:
      cont = 0;
      ret = -EINVAL;
      break;
    }
  }

  /* Verify argument combinations */
  if (ret) {
    fprintf(stderr, "Invalid arguments\n");
  } else if (!is_valid_powercap_control_type(control_type)) {
    fprintf(stderr, "Must specify control type NAME; value must not be empty or contain any '.' or '/' characters\n");
    ret = -EINVAL;
  } else if (!depth && (is_set_zone || is_set_constraint)) {
    fprintf(stderr, "Must specify -z/--zone with zone-level or constraint-level argument\n");
    ret = -EINVAL;
  } else if (depth && !(is_set_zone || is_set_constraint)) {
    fprintf(stderr, "Must specify zone-level or constraint-level argument with -z/--zone\n");
    ret = -EINVAL;
  } else if (!constraint.set && is_set_constraint) {
    fprintf(stderr, "Must specify -c/--constraint with constraint-level argument\n");
    ret = -EINVAL;
  } else if (constraint.set && !is_set_constraint) {
    fprintf(stderr, "Must specify constraint-level argument with -c/--constraint\n");
    ret = -EINVAL;
  } else if (!ct_enabled.set && !is_set_zone && !is_set_constraint) {
    fprintf(stderr, "Nothing to do\n");
    ret = -EINVAL;
  }
  if (ret) {
    print_usage();
    return EXIT_FAILURE;
  }

  /* Check if control type/zones/constraint exist */
  if (powercap_sysfs_control_type_exists(control_type)) {
    fprintf(stderr, "Control type does not exist\n");
    ret = -EINVAL;
  } else if (depth && powercap_sysfs_zone_exists(control_type, zones, depth)) {
    fprintf(stderr, "Zone does not exist\n");
    ret = -EINVAL;
  } else if (constraint.set && powercap_sysfs_constraint_exists(control_type, zones, depth, constraint.val)) {
    fprintf(stderr, "Constraint does not exist\n");
    ret = -EINVAL;
  }
  if (ret) {
    print_common_help();
    return EXIT_FAILURE;
  }

  /* Perform requested action(s) */
  if (ct_enabled.set) {
    c = powercap_sysfs_control_type_set_enabled(control_type, ct_enabled.val);
    if (c) {
      perror("Error setting control type enabled/disabled");
      ret |= c;
    }
  }
  if (reset_energy) {
    c = powercap_sysfs_zone_reset_energy_uj(control_type, zones, depth);
    if (c) {
      perror("Error setting zone energy counter");
      ret |= c;
    }
  }
  if (enabled.set) {
    c = powercap_sysfs_zone_set_enabled(control_type, zones, depth, enabled.val);
    if (c) {
      perror("Error setting zone enabled/disabled");
      ret |= c;
    }
  }
  if (power_limit.set) {
    c = powercap_sysfs_constraint_set_power_limit_uw(control_type, zones, depth, constraint.val, power_limit.val);
    if (c) {
      perror("Error setting constraint power limit");
      ret |= c;
    }
  }
  if (time_window.set) {
    c = powercap_sysfs_constraint_set_time_window_us(control_type, zones, depth, constraint.val, time_window.val);
    if (c) {
      perror("Error setting constraint time window");
      ret |= c;
    }
  }
  if (ret) {
    print_common_help();
  }

  return ret ? EXIT_FAILURE : EXIT_SUCCESS;
}
